Add require response option for BLE binary output

AshtonKem opened the pull request:
# What does this implement/fix? 

Was having an issue with my ESP32 not actually changing BLE characteristic values. Changing the write type from NO_RSP to RSP fixed it. Not entirely sure why.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
switch:
  - platform: template
    id: ashton_ooler_power
    name: "Ashton Ooler Power"
    turn_on_action: 
     - output.turn_on: ashton_ooler_power_output
    turn_off_action: 
      - output.turn_off: ashton_ooler_power_output
    lambda: |-
      if (id(ashton_ooler_power_internal).state > 0) {
        return true;
      } else {
        return false;
      }

output: 
  - platform: ble_client
    ble_client_id: ashton_ooler
    id: ashton_ooler_power_output
    service_uuid: 5C293993-D039-4225-92F6-31FA62101E96
    characteristic_uuid: 7A2623FF-BD92-4C13-BE9F-7023AA4ECB85


```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

AshtonKem committed:
Change write type for BLE binary output

buxtronix commented:
I *think* this should work as it's pretty standard in BLE to expect a response, and I confess I don't have a good reason as to why this was done with NO_RSP.
However we'd need to look carefully for any incoming bugs in case this breaks some devices.

AshtonKem commented:
Yeah, I didn't see any documentation that explained the difference, but then again I also didn't look very hard. In theory it shouldn't make a difference, but then again in theory it shouldn't have fixed my issue either. 🤷

jesserockz reviewed (changes requested):
This would be device specific and should become a config parameter

AshtonKem commented:
Seems reasonable, I’ll make that change. 

—
Ashton Kemerling

Sent from my iPhone. 

> On Jan 22, 2022, at 10:48 PM, Jesse Hills ***@***.***> wrote:
> 
> ﻿
> @jesserockz requested changes on this pull request.
> 
> This would be device specific and should become a config parameter
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you authored the thread.

AshtonKem committed:
New configuration parameter for ble_binary_output: response_required.

Controls whether or not the write type is ESP_GATT_WRITE_TYPE_NO_RSP or ESP_GATT_WRITE_TYPE_RSP.

Backwards compatible.

AshtonKem commented:
I'll update the docs once everyone is happy with the code changes.

AshtonKem committed:
Merge branch 'dev' into fix-ble-binary-output

AshtonKem committed:
Fix formatting

AshtonKem committed:
Fix formatting for C++ code

AshtonKem committed:
Fix C++ formatting again

jesserockz commented on esphome/components/ble_client/output/ble_binary_output.h:
```diff
@@ -25,9 +25,12 @@ class BLEBinaryOutput : public output::BinaryOutput, public BLEClientNode, publi
   void gattc_event_handler(esp_gattc_cb_event_t event, esp_gatt_if_t gattc_if,
                            esp_ble_gattc_cb_param_t *param) override;
+  void set_require_response(bool response) { this->require_response_ = response; }
+  bool is_response_required() const { return this->require_response_; }
```
I wouldn't worry about exposing a getter for this, and just use it as `this->require_response_` above in `write_state`
```suggestion
```

jesserockz commented on esphome/components/ble_client/output/ble_binary_output.cpp:
```diff
@@ -63,7 +63,11 @@ void BLEBinaryOutput::write_state(bool state) {
   uint8_t state_as_uint = (uint8_t) state;
   ESP_LOGV(TAG, "[%s] Write State: %d", this->char_uuid_.to_string().c_str(), state_as_uint);
-  chr->write_value(&state_as_uint, sizeof(state_as_uint));
+  if (this->is_response_required()) {
```
```suggestion
  if (this->require_response_) {
```

jesserockz commented on esphome/components/ble_client/output/__init__.py:
```diff
@@ -61,7 +63,8 @@ def to_code(config):
         )
         cg.add(var.set_char_uuid128(uuid128))
-
+    if CONF_REQUIRE_RESPONSE in config:
+        cg.add(var.set_require_response(config[CONF_REQUIRE_RESPONSE]))
```
Default is `False` so conditional not needed
```suggestion
    cg.add(var.set_require_response(config[CONF_REQUIRE_RESPONSE]))
```

jesserockz commented on esphome/components/ble_client/output/ble_binary_output.h:
```diff
@@ -25,9 +25,12 @@ class BLEBinaryOutput : public output::BinaryOutput, public BLEClientNode, publi
 
  protected:
   void write_state(bool state) override;
+  bool require_response_{false};
```
No default here as it is set in config validation/codegen
```suggestion
  bool require_response_;
```

AshtonKem committed:
Apply suggestions from @jesserockz.

jesserockz reviewed (approved):
LGTM. Just needs docs (or link in description if PR already exists)

@buxtronix ?

buxtronix commented:
LGTM also, pending docs update

AshtonKem commented:
https://github.com/esphome/esphome-docs/pull/1869
